// @subsquid/openreader/dist/test/setup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useServer = exports.useDatabase = exports.databaseDelete = exports.databaseInit = exports.db_config = exports.isCockroach = void 0;
const util_1 = require("@subsquid/util");
const gql_test_client_1 = require("gql-test-client");
const graphql_1 = require("graphql");
const pg_1 = require("pg");
const schema_1 = require("../gql/schema");
const server_1 = require("../server");
function isCockroach() {
    return process.env.DB_TYPE == 'cockroach';
}
exports.isCockroach = isCockroach;
exports.db_config = {
    host: 'localhost',
    port: parseInt((0, util_1.assertNotNull)(isCockroach() ? process.env.DB_PORT_COCKROACH : process.env.DB_PORT_PG)),
    user: 'root',
    password: 'root',
    database: 'defaultdb'
};
async function withClient(block) {
    let client = new pg_1.Client(exports.db_config);
    await client.connect();
    try {
        await block(client);
    }
    finally {
        await client.end();
    }
}
function databaseInit(sql) {
    return withClient(async (client) => {
        for (let i = 0; i < sql.length; i++) {
            await client.query(sql[i]);
        }
    });
}
exports.databaseInit = databaseInit;
function databaseDelete() {
    return withClient(async (client) => {
        await client.query(`DROP SCHEMA IF EXISTS root CASCADE`);
        await client.query(`CREATE SCHEMA root`);
    });
}
exports.databaseDelete = databaseDelete;
function useDatabase(sql) {
    before(async () => {
        await databaseDelete();
        await databaseInit(sql);
    });
}
exports.useDatabase = useDatabase;
function useServer(schema) {
    let client = new gql_test_client_1.Client('not defined');
    let db = new pg_1.Pool(exports.db_config);
    let info;
    before(async () => {
        info = await (0, server_1.serve)({
            db,
            model: (0, schema_1.buildModel)((0, schema_1.buildSchema)((0, graphql_1.parse)(schema))),
            port: 0,
            dialect: isCockroach() ? 'cockroach' : 'postgres'
        });
        client.endpoint = `http://localhost:${info.port}/graphql`;
    });
    after(() => info?.stop());
    after(() => db.end());
    return client;
}
exports.useServer = useServer;
