// @subsquid/cli/lib/commands/db/revert.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const typeorm_config_1 = require("@subsquid/typeorm-config");
const dotenv = __importStar(require("dotenv"));
const core_1 = require("@oclif/core");
const typeorm_1 = require("typeorm");
class RevertDb extends core_1.Command {
    async run() {
        dotenv.config();
        let cfg = {
            ...(0, typeorm_config_1.createOrmConfig)(),
            subscribers: [],
            synchronize: false,
            migrationsRun: false,
            dropSchema: false,
            logging: ["query", "error", "schema"]
        };
        let connection = await (0, typeorm_1.createConnection)(cfg);
        try {
            await connection.undoLastMigration({ transaction: 'all' });
        }
        finally {
            await connection.close().catch(err => null);
        }
    }
}
exports.default = RevertDb;
RevertDb.description = 'Revert the last performed migration';
