// @subsquid/util/lib/out.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.OutDir = exports.FileOutput = exports.Output = void 0;
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
class Output {
    constructor() {
        this.out = [];
        this.indent = '';
    }
    line(s) {
        if (s) {
            this.out.push(this.indent + s);
        }
        else {
            this.out.push('');
        }
    }
    block(start, cb) {
        this.line(start + ' {');
        this.indent += '  ';
        try {
            cb();
        }
        finally {
            this.indent = this.indent.slice(0, this.indent.length - 2);
        }
        this.line('}');
    }
    blockComment(lines) {
        if (!lines?.length)
            return;
        this.line(`/**`);
        lines.forEach((line) => this.line(' * ' + line)); // FIXME: escaping
        this.line(' */');
    }
    lazy(gen) {
        this.out.push({ indent: this.indent, gen });
    }
    toString() {
        let out = '';
        for (let i = 0; i < this.out.length; i++) {
            let line = this.out[i];
            if (typeof line == 'string') {
                out += line + '\n';
            }
            else {
                let lazy = line;
                lazy.gen().forEach(s => {
                    if (s) {
                        out += lazy.indent + s + '\n';
                    }
                    else {
                        out += '\n';
                    }
                });
            }
        }
        return out;
    }
}
exports.Output = Output;
class FileOutput extends Output {
    constructor(file) {
        super();
        this.file = file;
    }
    write() {
        fs_1.default.mkdirSync(path_1.default.dirname(this.file), { recursive: true });
        fs_1.default.writeFileSync(this.file, this.toString());
    }
}
exports.FileOutput = FileOutput;
class OutDir {
    constructor(dir) {
        this.dir = dir;
    }
    del() {
        fs_1.default.rmSync(this.dir, { recursive: true, force: true });
    }
    file(name) {
        return new FileOutput(this.path(name));
    }
    write(name, content) {
        const dst = this.path(name);
        fs_1.default.mkdirSync(path_1.default.dirname(dst), { recursive: true });
        fs_1.default.writeFileSync(dst, content);
    }
    child(name) {
        return new OutDir(this.path(name));
    }
    mkdir() {
        fs_1.default.mkdirSync(this.dir, { recursive: true });
    }
    add(name, srcFile) {
        let src = Array.isArray(srcFile) ? path_1.default.join(...srcFile) : srcFile;
        let dst = this.path(name);
        fs_1.default.mkdirSync(path_1.default.dirname(dst), { recursive: true });
        fs_1.default.copyFileSync(src, this.path(name));
    }
    path(name) {
        return path_1.default.join(this.dir, name);
    }
}
exports.OutDir = OutDir;
