// @subsquid/typeorm-config/lib/config.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.createOrmConfig = exports.createConnectionOptions = void 0;
const path = __importStar(require("path"));
const process = __importStar(require("process"));
const namingStrategy_1 = require("./namingStrategy");
function createConnectionOptions() {
    return {
        host: process.env.DB_HOST || 'localhost',
        port: process.env.DB_PORT ? parseInt(process.env.DB_PORT) : 5432,
        database: process.env.DB_NAME || 'postgres',
        username: process.env.DB_USER || 'postgres',
        password: process.env.DB_PASS || 'postgres'
    };
}
exports.createConnectionOptions = createConnectionOptions;
function createOrmConfig(options) {
    let dir = path.resolve(options?.projectDir || process.cwd());
    let model = resolveModel(path.join(dir, 'lib/model'));
    let migrationsDir = path.join(dir, 'db/migrations');
    return {
        type: 'postgres',
        namingStrategy: new namingStrategy_1.SnakeNamingStrategy(),
        entities: [model],
        migrations: [migrationsDir + '/*.js'],
        cli: {
            migrationsDir
        },
        ...createConnectionOptions()
    };
}
exports.createOrmConfig = createOrmConfig;
function resolveModel(model) {
    model = path.resolve(model || 'lib/model');
    try {
        return require.resolve(model);
    }
    catch (e) {
        throw new Error(`Failed to resolve model ${model}. Did you forget to run codegen or compile the code?`);
    }
}
