// @subsquid/cli/lib/commands/archive/rm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const baseUrl_1 = require("../../rest-client/baseUrl");
const request_1 = require("../../rest-client/request");
const creds_1 = require("../../creds");
class Remove extends core_1.Command {
    async run() {
        const { args: { name } } = await this.parse(Remove);
        const response = await (0, request_1.request)(`${baseUrl_1.baseUrl}/client/archives/${name}`, {
            method: 'delete',
            headers: {
                'Content-Type': 'application/json',
                authorization: `token ${(0, creds_1.getCreds)()}`,
            },
        });
        const responseBody = await response.json();
        if (response.status === 200) {
            this.log(`Deleted archive with name ${responseBody.name}`);
        }
    }
}
exports.default = Remove;
Remove.description = 'Delete an archive';
Remove.args = [
    {
        name: 'name',
        description: 'archive name',
        required: true,
    },
];
