// @subsquid/cli/lib/config/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getConfigField = exports.normalizeDefaults = exports.defaultApiUrl = exports.configFilePath = exports.configDirectory = void 0;
const fs_1 = require("fs");
const os_1 = require("os");
const path_1 = require("path");
exports.configDirectory = (0, path_1.resolve)((0, os_1.homedir)(), '.hydra-cli');
exports.configFilePath = (0, path_1.resolve)(exports.configDirectory, 'config.json');
exports.defaultApiUrl = 'https://saas.infra.gc.subsquid.io/api';
function writeDefaultConfigData() {
    const defaultConfigData = {
        apiUrl: exports.defaultApiUrl,
        credentials: 'empty'
    };
    (0, fs_1.writeFileSync)(exports.configFilePath, JSON.stringify(defaultConfigData), {
        flag: 'w',
        encoding: 'utf8',
    });
}
function normalizeDefaults() {
    if (!(0, fs_1.existsSync)(exports.configFilePath)) {
        if (!(0, fs_1.existsSync)(exports.configDirectory)) {
            (0, fs_1.mkdirSync)(exports.configDirectory);
        }
        writeDefaultConfigData();
        return;
    }
    let config;
    try {
        config = JSON.parse((0, fs_1.readFileSync)(exports.configFilePath, 'utf8'));
    }
    catch (e) {
        writeDefaultConfigData();
    }
}
exports.normalizeDefaults = normalizeDefaults;
function getConfigField(name) {
    normalizeDefaults();
    return JSON.parse((0, fs_1.readFileSync)(exports.configFilePath, 'utf8'))[name];
}
exports.getConfigField = getConfigField;
