// @subsquid/cli/lib/commands/squid/create.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const create_1 = require("../../rest-client/routes/create");
class Create extends core_1.Command {
    async run() {
        const { flags, args } = await this.parse(Create);
        const name = args.name;
        const description = flags.description;
        const logoUrl = flags.logo;
        const websiteUrl = flags.website;
        const createSquidMessage = await (0, create_1.create)(name, description, logoUrl, websiteUrl);
        this.log(createSquidMessage);
    }
}
exports.default = Create;
Create.description = 'Create a squid';
Create.args = [
    {
        name: 'name',
        description: 'squid name',
        required: true,
    },
];
Create.flags = {
    description: core_1.Flags.string({
        char: 'd',
        description: 'description',
        required: false,
    }),
    logo: core_1.Flags.string({
        char: 'l',
        description: 'logo url',
        required: false,
    }),
    website: core_1.Flags.string({
        char: 'w',
        description: 'website url',
        required: false,
    }),
};
