// @subsquid/cli/lib/commands/squid/ls.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const squids_1 = require("../../rest-client/routes/squids");
const versions_1 = require("../../rest-client/routes/versions");
class Ls extends core_1.Command {
    async run() {
        const { flags } = await this.parse(Ls);
        const noTruncate = !flags.truncate;
        const squidName = flags.name;
        if (squidName) {
            const deployments = await (0, versions_1.versionList)(squidName);
            if (deployments) {
                core_1.CliUx.ux.table(deployments, {
                    name: { header: 'version' },
                    artifactUrl: { header: 'endpoint' },
                    deploymentUrl: { header: 'source' },
                    status: {},
                    createdAt: { header: 'created at' },
                }, { 'no-truncate': noTruncate });
            }
        }
        else {
            const squids = await (0, squids_1.squidList)();
            if (squids) {
                core_1.CliUx.ux.table(squids, {
                    name: {},
                    description: {},
                }, { 'no-truncate': noTruncate });
            }
        }
    }
}
exports.default = Ls;
Ls.description = 'Squid or versions list';
Ls.flags = {
    name: core_1.Flags.string({
        char: 'n',
        description: 'squid name',
        required: false,
    }),
    truncate: core_1.Flags.boolean({
        char: 't',
        description: 'truncate data in columns: false by default',
        required: false,
        default: false,
    }),
};
