// @subsquid/cli/lib/commands/squid/tail.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const log_1 = require("../../rest-client/routes/log");
const utils_1 = require("../../utils");
class Tail extends core_1.Command {
    async run() {
        const { flags, args } = await this.parse(Tail);
        const nameAndVersion = args.nameAndVersion;
        const { squidName, versionName } = (0, utils_1.parseNameAndVersion)(nameAndVersion, this);
        const follow = flags.follow;
        const lines = flags.lines;
        this.log('Fetching logs...');
        await (0, log_1.log)(squidName, versionName, follow, lines);
    }
}
exports.default = Tail;
Tail.description = 'Getting logs about version';
Tail.args = [
    {
        name: 'nameAndVersion',
        description: 'name@version',
        required: true,
    },
];
Tail.flags = {
    follow: core_1.Flags.boolean({
        char: 'f',
        description: 'will continue streaming the new logs',
        required: false,
        default: false,
    }),
    lines: core_1.Flags.integer({
        char: 'l',
        description: 'output a specific number of lines (if "follow" is not set)',
        required: false,
        default: 50,
    }),
};
