// @subsquid/cli/lib/commands/squid/update.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@oclif/core");
const simple_git_1 = __importDefault(require("simple-git"));
const update_1 = require("../../rest-client/routes/update");
const utils_1 = require("../../utils");
const options = {
    baseDir: process.cwd(),
    binary: 'git',
};
const git = (0, simple_git_1.default)(options);
class Update extends core_1.Command {
    async run() {
        const { flags, args } = await this.parse(Update);
        const nameAndVersion = args.nameAndVersion;
        const { squidName, versionName } = (0, utils_1.parseNameAndVersion)(nameAndVersion, this);
        let deployUrl = flags.source;
        if (!deployUrl) {
            deployUrl = await (0, utils_1.buildRemoteUrlFromGit)(git, this);
        }
        else {
            deployUrl = deployUrl.split('#')[0].endsWith('.git')
                ? deployUrl
                : `${deployUrl.split('#')[0]}.git${deployUrl.split('#')[1]
                    ? '#' + deployUrl.split('#')[1]
                    : ''}`;
        }
        this.log(`🦑 Releasing the Squid at ${deployUrl}`);
        const result = await (0, update_1.update)(squidName, versionName, deployUrl, flags.hardReset);
        this.log('◷ You may now detach from the build process by pressing Ctrl + C. The Squid deployment will continue uninterrupted.');
        this.log('◷ The new squid will be available as soon as the deployment is complete.');
        await (0, utils_1.pollDeployPipelines)(squidName, versionName, result?.deploymentUrl || '', this);
        this.log('✔️ Done!');
    }
}
exports.default = Update;
Update.description = 'Update a version image';
Update.args = [
    {
        name: 'nameAndVersion',
        description: 'name@version',
        required: true,
    },
];
Update.flags = {
    source: core_1.Flags.string({
        char: 's',
        description: 'source',
        required: false,
    }),
    hardReset: core_1.Flags.boolean({
        char: 'r',
        description: 'perform a hard reset (db wipeout)',
        required: false,
        default: false
    })
};
